### DISCUSSION TITLE
Using Wix Custom Action (WixQuietExec) to install a driver

### DISCUSSION THREAD

**Discussion by bmtpolo** [wixtoolset/issues#8557] [Questions]
I'm trying to install a driver using pnputil and running into issues.  Here's a snippet of my WIX code

```
<Wix xmlns="http://wixtoolset.org/schemas/v4/wxs">
  <Fragment>
    <?define PNP_INSTALL = pnputil.exe /add-driver?>
    <?define PNP_UNINSTALL = pnputil.exe /delete-driver?>

    <SetProperty Id="UsbCAInstall" Value="&quot;$(PNP_INSTALL) [DriversFolder]usb_driver.inf /install&quot;" Before="UsbCAInstall" Sequence="execute" />
    <SetProperty Id="UsbCAUninstall" Value="&quot;$(PNP_UNINSTALL) [DriversFolder]usb_driver.inf&quot;" Before="UsbCAUninstall" Sequence="execute" />

    <CustomAction Id="UsbCAInstall" BinaryRef="Wix4UtilCA_$(sys.BUILDARCHSHORT)" DllEntry="WixQuietExec" Execute="deferred" Impersonate="no" Return="ignore"/>
    <CustomAction Id="UsbCAUninstall" BinaryRef="Wix4UtilCA_$(sys.BUILDARCHSHORT)" DllEntry="WixQuietExec" Execute="deferred" Impersonate="no" Return="ignore"/>

   <InstallExecuteSequence>
     <Custom Action="UsbCAInstall" Before="InstallFinalize" Condition="NOT Installed"/>
     <Custom Action="UsbCAUninstall" Before="InstallFinalize" Condition="(NOT UPGRADINGPRODUCTCOE) AND (REMOVE=&quot;ALL&quot;)"/>
    </InstallExecuteSequence>    
  </Fragment>
</Wix>
```

When I try to install the MSI, I get the following in the log
Action 11:07:52: UsbCAInstall. 
MSI (s) (78:B4) [11:07:52:188]: Executing op: CustomActionSchedule(Action=UsbCAInstall,ActionType=3137,Source=BinaryData,Target=WixQuietExec,CustomActionData="pnputil.exe /add-driver C:\Program Files\X_Drivers\usb_driver.inf /install")
MSI (s) (78:84) [11:07:52:188]: Invoking remote custom action. DLL: C:\WINDOWS\Installer\MSI956A.tmp, Entrypoint: WixQuietExec
MSI (s) (78:14) [11:07:52:188]: Generating random cookie.
MSI (s) (78:14) [11:07:52:204]: Created Custom Action Server with PID 31816 (0x7C48).
MSI (s) (78:18) [11:07:52:282]: Running as a service.
MSI (s) (78:18) [11:07:52:282]: Hello, I'm your 64bit Elevated Non-remapped custom action server.
WixQuietExec:  Error 0x80070002: Command failed to execute.
WixQuietExec:  Error 0x80070002: QuietExec Failed
WixQuietExec:  Error 0x80070002: Failed in ExecCommon method
CustomAction UsbCAInstall returned actual error code 1603 but will be translated to success due to continue marking

If anyone can point me to what I'm doing wrong, I would greatly appreciate it.

**Comment by wmanning** [comment]
Do you have a InstallExecuteSequence element?

Are you following this example in the WixToolset docs?:
https://wixtoolset.org/docs/tools/wixext/quietexec/#silent-running

**Reply by wmanning** [reply]
Looks like you have everything correct from the doc example except for the missing InstallExecuteSequence.

**Comment by bmtpolo** [comment]
I do have the install execute sequence (For some reason it's not showing in my original post

    <InstallExecuteSequence>
      <Custom Action="UsbCAInstall" Before="InstallFinalize" Condition="NOT Installed"/>
      <Custom Action="UsbCAUninstall" Before="InstallFinalize" Condition="(NOT UPGRADINGPRODUCTCOE) AND (REMOVE=&quot;ALL&quot;)"/>
    </InstallExecuteSequence>

**Reply by wmanning** [reply]
I see you would have had to since in the log it does should a valid path: CustomActionData="pnputil.exe /add-driver C:\Program Files\X_Drivers\usb_driver.inf /install".

Did you try running the installer as admin?
https://answers.microsoft.com/en-us/windows/forum/all/error-1925-and-1603-when-installing-what-do-i-do/65dd287e-8233-4fbc-9365-bdf4c1647b89

**Reply by wmanning** [reply]
actually looks like error code 1603 is a generic error: A fatal error occurred during installation...

From the example they don't have quotes around the entire command line.
Instead of

`Value="&quot;$(PNP_INSTALL) [DriversFolder]usb_driver.inf /install&quot;"`
try
`Value="$(PNP_INSTALL) &quot;[DriversFolder]usb_driver.inf&quot; /install"`
Technically don't need quotes around file path since no spaces but safety.

Maybe the starting quotes were messing things up.
